pkg/messaging/mqtt/pubsub.go

```diff
@@ -147,32 +152,30 @@ func (ps pubsub) Unsubscribe(id, topic string) error {
 
 	ok = token.WaitTimeout(ps.timeout)
 	if !ok {
-		return errUnsubscribeTimeout
+		return ErrUnsubscribeTimeout
 	}
 	if ok := s.delete(topic); !ok {
-		return errUnsubscribeDeleteTopic
+		return ErrUnsubscribeDeleteTopic
 	}
+	ps.subscriptions[id] = s
+
 	if len(s.topics) == 0 {
 		delete(ps.subscriptions, id)
 	}
 	return token.Error()
 }
 
-func (ps pubsub) mqttHandler(h messaging.MessageHandler) mqtt.MessageHandler {
-	return func(c mqtt.Client, m mqtt.Message) {
-		var msg messaging.Message
-		if err := proto.Unmarshal(m.Payload(), &msg); err != nil {
-			ps.logger.Warn(fmt.Sprintf("Failed to unmarshal received message: %s", err))
-			return
-		}
-		if err := h.Handle(msg); err != nil {
-			ps.logger.Warn(fmt.Sprintf("Failed to handle Mainflux message: %s", err))
-		}
-	}
-}
-
 func newClient(address, id string, timeout time.Duration) (mqtt.Client, error) {
-	opts := mqtt.NewClientOptions().SetUsername(username).AddBroker(address).SetClientID(id)
+	opts := mqtt.NewClientOptions().
+		SetUsername(username).
+		AddBroker(address).
+		SetClientID(id).
+		SetConnectionLostHandler(func(c mqtt.Client, err error) {
+			time.Sleep(500 * time.Millisecond)
+		}).
+		SetReconnectingHandler(func(c mqtt.Client, options *mqtt.ClientOptions) {
+			time.Sleep(500 * time.Millisecond)
+		})
```

Comment:
What's this?

Reply:
I implemented this when the mosquitto-broker was disconnecting sometimes, and causing tests to fail randomly.
This is just to give time for broker to reconnect, in case of failure while testing.
I found this on stack-overflow, and it fixed that bug of tests randomly failing. And 0.5 seconds is small enough that it doesn't cause tests to run for long durations.

Reply:
Did you try and test without them? Do we need both since effectively they do the same thing?

Reply:
Looks like the problem is in `TestAll` test. Please check the logic behind it.

Reply:
The problem is, when I subscribe to a topic using pubsub.Subscribe() then publish a message to that topic, using client.Publish() {where client is a basic mqtt-client from `"github.com/eclipse/paho.mqtt.golang"`} both don't work together.
That's why TestAll is failing.
I've removed that test, but still getting this error in TestSubscribe() now. I'm working on fixing this currently.